File: vipaq/src/Binacle.ViPaq/Compression/NoOpCodec.cs
namespace Binacle.ViPaq.Compression;

// Hands the bytes straight back. Not a wire codec: a blob with `Compressed = 1` must carry a real compressed
// stream (PROTOCOL.md §6).
//
// It makes `ProtocolEncoder` testable on its own. Set `Compressed = 1` and hand it this codec, and the
// compressed path runs end to end with the body still readable byte for byte - impossible through a real codec,
// because compressed bytes must never be compared (§6.1).
internal sealed class NoOpCodec : ICompressionCodec
{
	public byte[] Compress(ReadOnlySpan<byte> body)
	{
		return body.ToArray();
	}

	public byte[] Decompress(ReadOnlySpan<byte> compressed)
	{
		return compressed.ToArray();
	}
}

File: vipaq/packages/binacle-vipaq/src/compression/noOpCodec.ts
import {CompressionCodec} from "./compressionCodec";

// Ports C#: NoOpCodec. Hands the bytes straight back (a copy, so callers can't alias the source). Not a wire
// codec — no conforming blob is written with it, because Compressed = 1 must carry a real compressed stream
// (PROTOCOL.md §6). It exists to make ProtocolEncoder testable: set the compressed bit, hand it this codec, and
// the compressed path runs end to end with the framing and contents still readable byte for byte.
export const noOpCodec: CompressionCodec = {
	async compress(body: Uint8Array): Promise<Uint8Array<ArrayBuffer>> {
		return new Uint8Array(body);
	},

	async decompress(compressed: Uint8Array): Promise<Uint8Array<ArrayBuffer>> {
		return new Uint8Array(compressed);
	},
};
